TechnicalDocuments/Proposals/BabylonDesignLibrary.md

```diff
@@ -0,0 +1,196 @@
+
+## Introduction
+This proposal is about introducing a Design Library.
+
+## Motivation
+With our current infrastrure we have have reusable components which don't reuse the same style,
+we provide different styesheets based on the design.
+
+The idea behind this is to have reusable components with a predefined stylesheet displaying different data.
+The benefit of having a design library is
+1. to develop new UIs faster.
+2. to catch inconsistencies in the UI due to a deisgn flaw or due to an engineering mistake.
+
+The requirement is to cover most of the common use cases and *not* to extend to edge cases.
+Also we must reuse our current tools like `Bento`.
+
+For instance, in our codebase a common UI element is a rounded button with background color. 
+For that we use `Component.Button` and a stylesheet while this approach is functional is has a few drawbacks.
+The attributes of the button (background color, alpha, corner radius etc) can either 
+1. have been redefined in another part of the codebase which is good from a design point of view *and if* we search
+for them we can use them, *if we don't* search for them we will just reimplement something which already exists.
+2. or the attributes of the button haven't been redefined in another part of the application and we may not
+notice that so a design flaw will be implemented. We are humans after all :).
+
+
+## Proposed solution
+We can introduce `DesignLibrary` which will be responsible for creating these components following
+the semantics of the design styleguide which will be provided to us.
+
+The purpose of  `DesignLibrary` isn't to expose the full potential of `Bento` components but to 
+follow specific design guidelines.
+As an example of this is: 
+A `Bento` component supports different behaviours like  `Deletable` the goal of this
+proposal isn't to replicate them or to make a wrapper around them but to follow specific design guidelines
+and to offer reusable UI elements.
+
+Also with `DesignLibrary` we want to limit the exposure of specific components, in order to be able to introduce  
+drastic design changes in the future as easy as possible.
+For example `DesignLibrary.ComponentsBuilder.button` returns `AnyRenderable` instead of `Component.Button`
+because in the future we may decide that we need to use a different button component. We want to depend on abstractions.
+
+Furthermore, `DesignLibrary` can also host alerts, custom viewcontroller animations, etc.
+
+``` swift
+
+struct DesignLibrary {
+    let configuration: Configuration
+
+    var viewControllerBuilder: ViewControllerBuilder {
+        return ViewControllerBuilder(configuration: configuration)
+    }
+
+    var componentBuilder: ComponetsBuilder {
+        return ComponetsBuilder(configuration: configuration, traits: UITraitCollection())
+    }
+
+    init(
+        configuration: Configuration
+    ) {
+        self.configuration = configuration
+    }
+}
+
+extension DesignLibrary {
+    struct Configuration {
```

Comment:
It appears to me that `var traits: UITraitCollection` should be moved here, so that the config here can be updated with altered trait collections, and generate tokens _also_ based on the preferred content size category and (potentially) the user interface style (light/dark).

Reply:
and [nit] `Tokens`.

Reply:
if `var traits: UITraitCollection` moves there then it will make `DesignLibrary`depend into `UITraitCollection` which isn't idea since we may also add a `ViewControllerBuilder` in the future.

We can have support for multiple themes by modifying `Current.designLibrary` .
No?

Reply:
The question is more about how fonts (and maybe colors) are generated with regard to `UITraitCollection`, which has to live somewhere & has access to the current `UIContentSizeCategory`, if not through a `UITraitCollection` directly.

I don't quite understand how it is going to work with the current structure, unless you are proposing to ignore the trait collection provided by BentoKit `BoxViewController`, and just updating `Current.designLibrary` whenever the preferred content size category changed notification is posted.

Reply:
`Tokens` is a good name 👍